xvec/index.py

```diff
@@ -0,0 +1,193 @@
+    def crs(self) -> CRS:
+        """Returns the coordinate reference system of the index as a
+        :class:`pyproj.crs.CRS` object.
+        """
+        return self._crs
+
+    @property
+    def sindex(self) -> shapely.STRtree:
+        """Returns the spatial index, i.e., a :class:`shapely.STRtree` object.
+
+        It may build the index before returning it.
+        """
+        if self._sindex is None:
+            self._sindex = shapely.STRtree(self._index.index)
+        return self._sindex
+
+    @classmethod
+    def from_variables(
+        cls,
+        variables: Mapping[Any, Variable],
+        *,
+        options: Mapping[str, Any],
+    ):
+        # TODO: try getting CRS from coordinate attrs or GeometryArray
+        if "crs" not in options:
+            raise ValueError("a CRS must be provided")
+
+        index = PandasIndex.from_variables(variables, options={})
+        return cls(index, crs=options["crs"])
+
+    @classmethod
+    def concat(
+        cls,
+        indexes: Sequence[GeoVectorIndex],
+        dim: Hashable,
+        positions: Iterable[Iterable[int]] | None = None,
+    ) -> GeoVectorIndex:
+        crss = [idx.crs for idx in indexes]
+
+        if any([s != crss[0] for s in crss]):
+            raise ValueError("conflicting CRS for coordinates to concat")
+
+        indexes_ = [idx._index for idx in indexes]
+        index = PandasIndex.concat(indexes_, dim, positions)
+        return cls(index, crss[0])
+
+    def create_variables(
+        self, variables: Mapping[Any, Variable] | None = None
+    ) -> dict[Hashable, Variable]:
+        return self._index.create_variables(variables)
+
+    def to_pandas_index(self) -> pd.Index:
+        return self._index.index
+
+    def isel(self, indexers: Mapping[Any, Any]):
+        index = self._index.isel(indexers)
+
+        if index is not None:
+            return type(self)(index, self.crs)
+        else:
+            return None
+
+    def _sel_sindex(self, labels, method, tolerance):
+        # only one coordinate supported
+        assert len(labels) == 1
+        label = next(iter(labels.values()))
+
+        if method != "nearest":
+            if not isinstance(label, shapely.Geometry):
+                raise ValueError(
+                    "selection with another method than nearest only supports "
+                    "a single geometry as input label."
+                )
+
+        if isinstance(label, DataArray):
+            label_array = label._variable._data
+        elif isinstance(label, Variable):
+            label_array = label._data
+        elif isinstance(label, shapely.Geometry):
+            label_array = np.array([label])
+        else:
+            label_array = np.array(label)
+
+        # check for possible CRS of geometry labels
+        # (by default assume same CRS than the index)
+        if hasattr(label_array, "crs") and label_array.crs != self.crs:
+            raise ValueError("conflicting CRS for input geometries")
```

Comment:
This will also need a change if crs becomes optional.

Reply:
What should we do when label CRS is not None and index CRS is None? Raise or ignore the label CRS?

Reply:
In geopandas we warn that the two CRS are not matching but let the operation happen. I'd do the same.

Reply:
When two CRS are checked against each other like in overlay or sjoin, if one gdf has a CRS and the other does not, it always warns and executes.

Reply:
Hmm I wonder if we shouldn't have a special case for `.sel()`. I find it convenient to just pass one or more (list/array) shapely geometries as input labels without an explicit CRS. A warning may be annoying.